sky/serve/autoscalers.py

```diff
@@ -238,3 +240,261 @@ def _get_replica_ids_to_scale_down(num_limit: int) -> List[int]:
+        )
+        print(
+            f'autoscaler.collect_request_information(self.total_request_in_window): {self.total_request_in_window}'
+        )
+        print(
+            f'autoscaler.collect_request_information(self.request_distribution): {self.request_distribution}'
+        )
+
+    def _get_accelerator_override_dict(
+            self, instance_type: AcceleratorType) -> Dict[str, Any]:
+        return {'accelerators': f'{instance_type.value}:1'}
+
+    def _get_autoscaler_decision(
+            self,
+            operator: AutoscalerDecisionOperator,
+            accelerator: Optional[AcceleratorType] = None,
+            is_primary: Optional[bool] = None,
+            replica_id: Optional[int] = None) -> AutoscalerDecision:
+        if operator == AutoscalerDecisionOperator.SCALE_UP:
+            assert accelerator, accelerator
+            assert is_primary, is_primary
+            operator_target = self._get_accelerator_override_dict(accelerator)
+            operator_target.update({
+                'is_primary': True if is_primary else False,
+                'is_fallback': False if is_primary else True,
+            })
+            decision = AutoscalerDecision(operator, target=operator_target)
+        else:  # SCALE_DOWN
+            assert replica_id, replica_id
+            decision = AutoscalerDecision(operator, target=replica_id)
+        return decision
+
+    def _get_fallback_allocation(self, accelerator: AcceleratorType):
+        if accelerator == AcceleratorType.A10:
+            num, fallback_type = 0, None
+        elif accelerator == AcceleratorType.A100:
+            num, fallback_type = 4, AcceleratorType.A10
+        return num, fallback_type
+
+    def fallback_scale_down_sync(
+        self,
+        service_name: str,
+        replica_manager: 'replica_managers.ReplicaManager',
+    ) -> None:
+        replica_infos = serve_state.get_replica_infos(service_name)
+        replica_info_dicts = [
+            info.to_info_dict(
+                with_handle=env_options.Options.SHOW_DEBUG_INFO.get())
+            for info in replica_infos
+        ]
+        logger.info(
+            f'All replica info before fallback sync: {replica_info_dicts}')
+        # Iterate through replica_infos and check for primary replicas
+        # Check if each primary replica in state of ReplicaStatus.READY
+        # has existing fallback replicas
+        ready_primary_replica_infos = [
+            info for info in replica_infos if info.is_primary and
+            info.is_ready and info.fallback_replica_id_list
+        ]
+        for info in ready_primary_replica_infos:
+            fallback_replica_id_to_terminate = info.fallback_replica_id_list
+            for replica_id in fallback_replica_id_to_terminate:
+                replica_manager.scale_down(replica_id)
+                info.fallback_replica_id_list.remove(replica_id)
+                serve_state.add_or_update_replica(service_name, info.replica_id,
```

Comment:
I'm trying to understand this line. Why does fallback_scale_down_sync add the primary replica to serve_state for every fallback replica that is spun down?

Reply:
Oh nvm, it's to *update* not add. Since the fallback_replica_id_list is updated. Ignore me :)